src/lib/pages/frontsite/PatronProfile/PatronCurrentLoans/BrwReqLoanExtendButton.js

```diff
@@ -0,0 +1,215 @@
+};
+
+const REQUEST_MESSAGES = {
+  IS_PENDING: requestDate =>
+    `Extension pending since ${requestDate.toLocaleString({
+      month: 'long',
+      day: 'numeric',
+    })}`,
+  IS_DECLINED: 'Extension declined by the library',
+};
+
+class BorrowingRequestFetcher {
+  get = async pid => {
+    this.cancellablePromise = withCancel(borrowingRequestApi.get(pid));
+    const resp = await this.cancellablePromise.promise;
+    return resp.data;
+  };
+
+  cancel = () => {
+    this.cancellablePromise && this.cancellablePromise.cancel();
+  };
+}
+
+class ExtensionRequester {
+  doRequest = async brwReqPid => {
+    this.cancellablePromise = withCancel(
+      borrowingRequestApi.requestExtension(brwReqPid)
+    );
+    const resp = await this.cancellablePromise.promise;
+    return resp.data;
+  };
+
+  cancel = () => {
+    this.cancellablePromise && this.cancellablePromise.cancel();
+  };
+}
+
+const validateLoanExtension = brwReqLoan => {
+  let isValid = true;
+  let msg = null;
+
+  const isTooEarly =
+    brwReqLoan.metadata.end_date.diffNow('days').days > CONFIG_EXPIRE_DAYS;
+  if (isTooEarly) {
+    isValid = false;
+    msg = INFO_MESSAGES.IS_TOO_EARLY(brwReqLoan.metadata.end_date);
+  }
+
+  const maxExtensionsReached =
+    _get(brwReqLoan, 'metadata.extension_count', 0) >
+    invenioConfig.circulation.extensionsMaxCount;
+  if (maxExtensionsReached) {
+    isValid = false;
+    msg = INFO_MESSAGES.MAX_EXTENSIONS;
+  }
+
+  return {
+    isValid: isValid,
+    cmp: msg ? (
+      <Popup
+        content={msg}
+        trigger={<Icon name="info" />}
+        position="top right"
+      />
+    ) : null,
+  };
+};
+
+const validateExtensionRequest = extension => {
+  let isValid = true;
+  let msg = null;
+
+  if (!_isEmpty(extension)) {
+    const extensionStatus = extension.status;
+    const isPending = extensionStatus === 'PENDING';
+    if (isPending) {
+      isValid = false;
+      const requestDate = extension.request_date;
+      msg = REQUEST_MESSAGES.IS_PENDING(requestDate);
+    }
+
+    const isDeclined = extensionStatus === 'DECLINED';
+    if (isDeclined) {
+      isValid = false;
+      msg = REQUEST_MESSAGES.IS_DECLINED;
+    }
+  }
+
+  return {
+    isValid: isValid,
+    cmp: msg ? (
+      <>
+        <span>{msg}</span>{' '}
+      </>
+    ) : null,
+  };
+};
+
+class BrwReqLoanExtendButton extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      requestExtensionIsLoading: false,
+      brwReqDetailsIsLoading: false,
+      brwReqDetails: {},
+      brwReqDetailsError: {},
+    };
+    this.borrowingRequestFetcher = new BorrowingRequestFetcher();
+    this.extensionRequester = new ExtensionRequester();
+  }
+
+  componentDidMount() {
+    const { brwReqLoan } = this.props;
+    const brwReqPid = brwReqLoan.metadata.item_pid.value;
+    this.fetchBorrowingRequestDetails(brwReqPid);
+  }
+
+  componentWillUnmount() {
+    this.borrowingRequestFetcher.cancel();
```

Comment:
the `.cancel` is a bit misleading as a name, my first thought was you are canceling the borrowing request here
In the `Promise` it is named `reject`, what about `rejectPromise()` to be very clear? can be done later on, because I see `withCancel` function could follow the naming too, maybe too many changes for now

Reply:
well, it is kind of cancelling!
The idea is that you should not expose internals of `withCancel`. What you want to do here is the following: when you unmount a component please cancel the HTTP request, otherwise you have a memory leak.
Then, it is an internal thingy that the `Promise` is a rejected: if tomorrow we change the implementation of `withCancel` and we don't use Promises any more (made up example), then you don't want to change reject to something else.
The point is: hide internals, expose a clear interface.

What is bothering me here is the `.promise` here:`await this.cancellableCancelRequest.promise;`. It would be great just to have `await this.cancellableCancelRequest;`

Reply:
:+1: I understand I just try to separate the action names from the code specific ones. I guess good for now, since we know the memory leak issue.